antenna/app.py

```diff
@@ -144,107 +50,114 @@ def build_config_manager():
+                "used instead."
+            ),
+        )
 
     def __init__(self, config):
+        super().__init__()
         self.config_manager = config
         self.config = config.with_options(self)
+        self._all_resources = {}
 
-    def __call__(self, key):
-        """Return configuration for given key."""
-        return self.config(key)
+        # This is the crashmover which takes a crash report, saves it, and publishes
+        # it
+        self.crashmover = CrashMover(config.with_namespace("crashmover"))
 
+        # This is the breakpad resource that handles incoming crash reports POSTed to
+        # /submit
+        self.breakpad = BreakpadSubmitterResource(
+            config=config.with_namespace("breakpad"), crashmover=self.crashmover
+        )
 
-class AntennaAPI(falcon.API):
-    """Falcon API for Antenna."""
+        self.hb = HeartbeatManager()
 
-    def __init__(self, config):
-        super().__init__()
-        self.config = config
-        self._all_resources = {}
+    def get_components(self):
+        """Return map of namespace -> component for traversing component tree
 
-        self.hb = HeartbeatManager(config)
+        This is only used for logging configuration for components that
+        are configurable at runtime using Everett.
+
+        """
+        return {
+            "crashmover": self.crashmover,
+            "breakpad": self.breakpad,
+        }
+
+    def setup(self):
+        # Set up logging and sentry first, so we have something to log to. Then
+        # build and log everything else.
+        setup_logging(
+            logging_level=self.config("logging_level"),
+            debug=self.config("local_dev_env"),
+            host_id=self.config("host_id"),
+            processname="antenna",
+        )
+
+        # Set a Sentry client if we're so configured
+        set_sentry_client(self.config("secret_sentry_dsn"), self.config("basedir"))
+
+        # Set up Sentry exception logger if we're so configured
+        setup_sentry_logging()
+
+        # Set up metrics
+        setup_metrics(
+            statsd_host=self.config("statsd_host"),
+            statsd_port=self.config("statsd_port"),
+            statsd_namespace=self.config("statsd_namespace"),
+            debug=self.config("local_dev_env"),
+        )
+
+        # Set up Dockerflow and health-related routes
+        self.add_route(
+            "version",
+            "/__version__",
+            VersionResource(basedir=self.config("basedir")),
+        )
+        self.add_route("heartbeat", "/__heartbeat__", HeartbeatResource(app=self))
+        self.add_route("lbheartbeat", "/__lbheartbeat__", LBHeartbeatResource())
+        self.add_route("broken", "/__broken__", BrokenResource())
+
+        # Set up breakpad submission route
+        self.add_route("breakpad", "/submit", self.breakpad)
+
+        # Log runtime configuration
+        log_config(LOGGER, self.config_manager, self)
```

Comment:
The route stuff is now part of the app code rather than part of the `get_app()` code.